// ngx-custom-validators/esm2015/src/app/included-in/directive.js
import { Directive, Input, forwardRef } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
import { includedIn } from './validator';
import * as ɵngcc0 from '@angular/core';
const INCLUDED_IN_VALIDATOR = {
    provide: NG_VALIDATORS,
    useExisting: forwardRef(() => IncludedInValidator),
    multi: true
};
export class IncludedInValidator {
    ngOnInit() {
        this.validator = includedIn(this.includedIn);
    }
    ngOnChanges(changes) {
        for (const key in changes) {
            if (key === 'includedIn') {
                this.validator = includedIn(changes[key].currentValue);
                if (this.onChange) {
                    this.onChange();
                }
            }
        }
    }
    validate(c) {
        return this.validator(c);
    }
    registerOnValidatorChange(fn) {
        this.onChange = fn;
    }
}
IncludedInValidator.ɵfac = function IncludedInValidator_Factory(t) { return new (t || IncludedInValidator)(); };
IncludedInValidator.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: IncludedInValidator, selectors: [["", "includedIn", "", "formControlName", ""], ["", "includedIn", "", "formControl", ""], ["", "includedIn", "", "ngModel", ""]], inputs: { includedIn: "includedIn" }, features: [ɵngcc0.ɵɵProvidersFeature([INCLUDED_IN_VALIDATOR]), ɵngcc0.ɵɵNgOnChangesFeature] });
IncludedInValidator.propDecorators = {
    includedIn: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(IncludedInValidator, [{
        type: Directive,
        args: [{
                selector: '[includedIn][formControlName],[includedIn][formControl],[includedIn][ngModel]',
                providers: [INCLUDED_IN_VALIDATOR]
            }]
    }], null, { includedIn: [{
            type: Input
        }] }); })();
