// ngx-custom-validators/esm2015/src/app/not-matching/directive.js
import { Directive, Input, forwardRef } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
import { notMatching } from './validator';
import * as ɵngcc0 from '@angular/core';
const NOT_MATCHING_VALIDATOR = {
    provide: NG_VALIDATORS,
    useExisting: forwardRef(() => NotMatchingValidator),
    multi: true
};
export class NotMatchingValidator {
    ngOnInit() {
        this.validator = notMatching(this.notMatching);
    }
    ngOnChanges(changes) {
        for (const key in changes) {
            if (key === 'notMatching') {
                this.validator = notMatching(changes[key].currentValue);
                if (this.onChange) {
                    this.onChange();
                }
            }
        }
    }
    validate(c) {
        return this.validator(c);
    }
    registerOnValidatorChange(fn) {
        this.onChange = fn;
    }
}
NotMatchingValidator.ɵfac = function NotMatchingValidator_Factory(t) { return new (t || NotMatchingValidator)(); };
NotMatchingValidator.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: NotMatchingValidator, selectors: [["", "notMatching", "", "formControlName", ""], ["", "notMatching", "", "formControl", ""], ["", "notMatching", "", "ngModel", ""]], inputs: { notMatching: "notMatching" }, features: [ɵngcc0.ɵɵProvidersFeature([NOT_MATCHING_VALIDATOR]), ɵngcc0.ɵɵNgOnChangesFeature] });
NotMatchingValidator.propDecorators = {
    notMatching: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NotMatchingValidator, [{
        type: Directive,
        args: [{
                selector: '[notMatching][formControlName],[notMatching][formControl],[notMatching][ngModel]',
                providers: [NOT_MATCHING_VALIDATOR]
            }]
    }], null, { notMatching: [{
            type: Input
        }] }); })();
