// ngx-custom-validators/esm2015/src/app/not-included-in/directive.js
import { Directive, Input, forwardRef } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
import { notIncludedIn } from './validator';
import * as ɵngcc0 from '@angular/core';
const NOT_INCLUDED_IN_VALIDATOR = {
    provide: NG_VALIDATORS,
    useExisting: forwardRef(() => NotIncludedInValidator),
    multi: true
};
export class NotIncludedInValidator {
    ngOnInit() {
        this.validator = notIncludedIn(this.notIncludedIn);
    }
    ngOnChanges(changes) {
        for (const key in changes) {
            if (key === 'notIncludedIn') {
                this.validator = notIncludedIn(changes[key].currentValue);
                if (this.onChange) {
                    this.onChange();
                }
            }
        }
    }
    validate(c) {
        return this.validator(c);
    }
    registerOnValidatorChange(fn) {
        this.onChange = fn;
    }
}
NotIncludedInValidator.ɵfac = function NotIncludedInValidator_Factory(t) { return new (t || NotIncludedInValidator)(); };
NotIncludedInValidator.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: NotIncludedInValidator, selectors: [["", "notIncludedIn", "", "formControlName", ""], ["", "notIncludedIn", "", "formControl", ""], ["", "notIncludedIn", "", "ngModel", ""]], inputs: { notIncludedIn: "notIncludedIn" }, features: [ɵngcc0.ɵɵProvidersFeature([NOT_INCLUDED_IN_VALIDATOR]), ɵngcc0.ɵɵNgOnChangesFeature] });
NotIncludedInValidator.propDecorators = {
    notIncludedIn: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NotIncludedInValidator, [{
        type: Directive,
        args: [{
                selector: '[notIncludedIn][formControlName],[notIncludedIn][formControl],[notIncludedIn][ngModel]',
                providers: [NOT_INCLUDED_IN_VALIDATOR]
            }]
    }], null, { notIncludedIn: [{
            type: Input
        }] }); })();
