// ng2-currency-mask/esm2015/lib/currency-mask.module.js
import { CommonModule } from "@angular/common";
import { NgModule } from "@angular/core";
import { FormsModule } from "@angular/forms";
import { CurrencyMaskDirective } from "./currency-mask.directive";
import * as ɵngcc0 from '@angular/core';
export class CurrencyMaskModule {
}
CurrencyMaskModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: CurrencyMaskModule });
CurrencyMaskModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function CurrencyMaskModule_Factory(t) { return new (t || CurrencyMaskModule)(); }, imports: [[
            CommonModule,
            FormsModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(CurrencyMaskModule, { declarations: function () { return [CurrencyMaskDirective]; }, imports: function () { return [CommonModule,
        FormsModule]; }, exports: function () { return [CurrencyMaskDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(CurrencyMaskModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    FormsModule
                ],
                declarations: [
                    CurrencyMaskDirective
                ],
                exports: [
                    CurrencyMaskDirective
                ]
            }]
    }], null, null); })();
