// angular-cropperjs/esm5/lib/angular-cropperjs.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CropperComponent } from './cropper/cropper.component';
import { CommonModule } from '@angular/common';
import * as ɵngcc0 from '@angular/core';
var AngularCropperjsModule = /** @class */ (function () {
    function AngularCropperjsModule() {
    }
AngularCropperjsModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: AngularCropperjsModule });
AngularCropperjsModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function AngularCropperjsModule_Factory(t) { return new (t || AngularCropperjsModule)(); }, imports: [[
            CommonModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(AngularCropperjsModule, { declarations: function () { return [CropperComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [CropperComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AngularCropperjsModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule
                ],
                declarations: [CropperComponent],
                exports: [CropperComponent]
            }]
    }], function () { return []; }, null); })();
    return AngularCropperjsModule;
}());
export { AngularCropperjsModule };
